libdrgn/program.c

```diff
@@ -1075,6 +1116,31 @@ drgn_thread_iterator_init_linux_kernel(struct drgn_thread_iterator *it)
 	return NULL;
 }
 
+static struct drgn_error *
+drgn_thread_iterator_init_gdbremote(struct drgn_thread_iterator *it)
+{
+	struct drgn_program *prog = it->prog;
+	struct drgn_thread thread = {
+	    .prog = prog,
+	    // Until we implement query packet parsing in the gdbremote code
+	    // then we are only able to debug the stopped thread.
+	    .tid = 1,
+	};
+
+	prog->main_thread =
+	    drgn_thread_set_search(&prog->thread_set, &thread.tid).entry;
```

Comment:
I don't think the thread set is the best fit for this, since the available threads are dynamic based on whatever the remote server tells us, right?

Another thing I'm curious about, does the remote protocol have a notion of a main or current thread? From what I can tell, you're free to start and stop threads at will via protocol commands, right?

(I'm assuming you just did it this way to make it easy for the minimal initial support, so I'm mainly asking to think about how the drgn API may need to be extended to represent remote protocol concepts.)

Reply:
> I don't think the thread set is the best fit for this, since the available threads are dynamic based on whatever the remote server tells us, right?

I think it would only be dynamic if we put the remote side into non-stop mode. Otherwise when we are talking to the remote then the whole system is stopped. In that case I figured using the thread_set would allow us to cache things (which will probably be desirable on a 115200 baud link). Having said that if we needed to cache more aggressively (including agressive lazyiness) we would probably need a different approach.



> Another thing I'm curious about, does the remote protocol have a notion of a main or current thread? From what I can tell, you're free to start and stop threads at will via protocol commands, right?

Likewise, that can happen in non-stop mode but I'd suggest the initial development focus should be for all-stop mode, if only because that "feels" much like a core dump, only with specialist memory and register options.